pkg/helm/chartproxy/repos.go

```diff
@@ -0,0 +1,224 @@
+package chartproxy
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sigs.k8s.io/yaml"
+	"strings"
+
+	"helm.sh/helm/v3/pkg/repo"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/openshift/library-go/pkg/crypto"
+)
+
+var (
+	helmChartRepositoryGVK = schema.GroupVersionResource{
+		Group:    "helm.openshift.io",
+		Version:  "v1beta1",
+		Resource: "helmchartrepositories",
+	}
+)
+
+const (
+	configNamespace = "openshift-config"
+)
+
+type helmRepo struct {
+	Name            string
+	Url             *url.URL
+	TlsClientConfig *tls.Config
+}
+
+type TLSConfigGetter interface {
+	Get() (*tls.Config, error)
+}
+
+func (repo helmRepo) Get() (*tls.Config, error) {
+	return repo.TlsClientConfig, nil
+}
+
+func (repo helmRepo) httpClient() (*http.Client, error) {
+	tlsConfig, err := repo.Get()
+	if err != nil {
+		return nil, err
+	}
+	tr := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	client := &http.Client{Transport: tr}
+	return client, nil
+}
+
+func (hr helmRepo) IndexFile() (*repo.IndexFile, error) {
+	var indexFile repo.IndexFile
+	httpClient, err := hr.httpClient()
+	if err != nil {
+		return nil, err
+	}
+	indexUrl := hr.Url.String()
+	if !strings.HasSuffix(indexUrl, "/index.yaml") {
+		indexUrl += "/index.yaml"
+	}
+	resp, err := httpClient.Get(indexUrl)
+	if err != nil {
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(body, &indexFile)
+	if err != nil {
+		return nil, err
+	}
+	return &indexFile, nil
+}
+
+type HelmRepoGetter interface {
+	List() ([]*helmRepo, error)
+}
+
+type helmRepoGetter struct {
+	Client     dynamic.Interface
+	CoreClient corev1.CoreV1Interface
+}
+
+func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmRepo, error) {
+	h := &helmRepo{}
+	urlValue, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "url")
```

Comment:
If you bump openshift/api, you should have a real struct with these fields, which would be better for type safety.

Reply:
We would love too, and we tried to  add `openshift/client-go` as the dependency (it contains all necessary types), but we cannot make it work currently in console, because client-go depends on k8s 1.19, but Helm (we depend on here) requires k8s 1.18. Tried to force it, but Helm is not yet compatible to k8s 1.19.

Reply:
is there any plan to make it compatible ?

Reply:
sure, when Helm moves to 1.19.